lib/heimdall_tools/nikto_mapper.rb

```diff
@@ -0,0 +1,152 @@
+    e.yield '\\'
+  end
+end
+
+module HeimdallTools
+  class NiktoMapper
+    def initialize(nikto_json, name=nil, verbose = false)
+      @nikto_json = nikto_json
+      @verbose = verbose
+
+      begin
+        @nikto_nist_mapping = parse_mapper
+      rescue StandardError => e
+        raise "Invalid Nikto to NIST mapping file: Exception: #{e}"
+      end
+
+        # TODO: Support Multi-target scan results
+        # Nikto multi-target scans generate invalid format JSONs
+        # Possible workaround to use https://stackoverflow.com/a/58209963/1670307
+
+      begin
+        @project = JSON.parse(nikto_json)
+      rescue StandardError => e
+        raise "Invalid Nikto JSON file provided Exception: #{e}"
+      end
+    end
+
+    def extract_scaninfo(project)
+      info = {}
+      begin
+        info['policy'] = 'Nikto Website Scanner'
+        info['version'] = NA_STRING
+        info['projectName'] = "Host: #{project['host']} Port: #{project['port']}"
+        info['summary'] = "Banner: #{project['banner']}"
+
+        info
+      rescue StandardError => e
+        raise "Error extracting project info from nikto JSON file provided Exception: #{e}"
+      end
+    end
+
+    def finding(vulnerability)
+      finding = {}
+      finding['status'] = 'failed'
+      finding['code_desc'] = "URL : #{vulnerability['url'].to_s } Method: #{vulnerability['method'].to_s}"
+      finding['run_time'] = NA_FLOAT
+
+      # Nikto results JSON does not profile scan timestamp; using current time to satisfy HDF format
+      finding['start_time'] = NA_STRING
+      [finding]
+    end
+
+    def nist_tag(niktoid)
+      entries = @nikto_nist_mapping.select { |x| niktoid.eql?(x[:niktoid].to_s) }
+      tags = entries.map { |x| x[:nistid] }
+      tags.empty? ? DEFAULT_NIST_TAG : tags.flatten.uniq
+    end
+
+    def impact(severity)
+      IMPACT_MAPPING[severity.to_sym]
+    end
+
+    def parse_mapper
+      csv_data = CSV.read(NIKTO_NIST_MAPPING_FILE, **{ encoding: 'UTF-8',
+                                                   headers: true,
+                                                   header_converters: :symbol})
+      csv_data.map(&:to_hash)
+    end
+
+    def desc_tags(data, label)
+      { "data": data || NA_STRING, "label": label || NA_STRING }
+    end
+
+    # Nikto report could have multiple vulnerability entries for multiple findings of same issue type.
+    # The meta data is identical across entries 
+    # method collapse_duplicates return unique controls with applicable findings collapsed into it.
+    def collapse_duplicates(controls)
+      unique_controls = []
+
+      controls.map { |x| x['id'] }.uniq.each do |id|
+        collapsed_results = controls.select { |x| x['id'].eql?(id) }.map {|x| x['results']}
+        unique_control = controls.find { |x| x['id'].eql?(id) }
+        unique_control['results'] = collapsed_results.flatten
+        unique_controls << unique_control
+      end
+      unique_controls
+    end
```

Comment:
This function seems to just be copy pasted between all of the mappers. It would be extremely easy to just move this up to a parent class and make all the mappers inherit from that class. (Similar to how Rails has ApplicationRecord in place of ActiveRecord as the base for all your classes by default).

Reply:
Method moved to a centralized Utils module